src/app/users/editServices.js

```diff
@@ -62,12 +67,26 @@ const post = async (req, res) => {
     return res.render('users/views/editServices', model);
   }
 
+  saveRoleInSession(req,selectedRoles);
+
+  return res.redirect(`${req.params.sid}/confirm-edit-service`)
+};
+
+const saveRoleInSession = (req, selectedRoles) => {
   req.session.service = {
     roles: selectedRoles
   };
-  return res.redirect(`${req.params.sid}/confirm-edit-service`)
 };
 
+const haveRolesBeenUpdated= (req, selectedRoles) => {
+  if(req.session.service 
+    && req.session.service.roles  
+    && req.session.service.roles.length === selectedRoles.length){
```

Comment:
You are simply checking the number of roles selected matches the current number of roles, this means I could change the  role I have selected to a different one but I wouldn't see the review changes page if the length was the same.

You need to check for the items in the array selectedRoles to see if they match what is in the req.session.service.roles array

Reply:
Now doing deep comparison of the session roles and selected roles